### PR TITLE
[Config] finish adding parameter types

### PR DESCRIPTION
**Pull request by Tobion**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | yno
| New feature?  | no 
| BC breaks?    | no     
| Deprecations? | no 
| Tests pass?   | yes    
| Fixed tickets | #32179
| License       | MIT
| Doc PR        | symfony/symfony-docs#... 

Completes some missing things in #32201

### CODE REVIEW

File: src/Symfony/Bundle/FrameworkBundle/Routing/DelegatingLoader.php [hunk 1]
```diff
@@ -40,7 +40,7 @@ public function __construct(LoaderResolverInterface $resolver, array $defaultOpt
     /**
      * {@inheritdoc}
      */
-    public function load($resource, $type = null)
+    public function load($resource, string $type = null)
```
Comment (derrabus) [inline src/Symfony/Bundle/FrameworkBundle/Routing/DelegatingLoader.php]: This will break compatibility with Config 4.4. Please either revert this change or update FrameworkBundle‘s composer.json accordingly.

File: src/Symfony/Component/DependencyInjection/Config/ContainerParametersResourceChecker.php [hunk 2]
```diff
@@ -39,7 +39,7 @@ public function supports(ResourceInterface $metadata)
     /**
      * {@inheritdoc}
      */
-    public function isFresh(ResourceInterface $resource, $timestamp)
+    public function isFresh(ResourceInterface $resource, int $timestamp)
```
Comment (derrabus) [inline src/Symfony/Component/DependencyInjection/Config/ContainerParametersResourceChecker.php]: This will break compatibility with Config 4.4. Please either revert this change or update DependencyInjection‘s composer.json accordingly.

File: src/Symfony/Component/Routing/Loader/AnnotationClassLoader.php [hunk 3]
```diff
@@ -94,7 +94,7 @@ public function setRouteAnnotationClass($class)
      *
      * @throws \InvalidArgumentException When route can't be parsed
      */
-    public function load($class, $type = null)
+    public function load($class, string $type = null)
```
Comment (derrabus) [inline src/Symfony/Component/Routing/Loader/AnnotationClassLoader.php]: This will break compatibility with Config 4.4. Please either revert this change or update Routing‘s composer.json accordingly.

**Review by derrabus — CHANGES_REQUESTED** [review]
Your changes make certain other components in version 5 incompatible with Config 4.4. Im not sure if we should break that compatibility just for adding type-hints.

File: src/Symfony/Bundle/FrameworkBundle/Routing/DelegatingLoader.php (hunk 1 above)
Comment (Tobion) [reply src/Symfony/Bundle/FrameworkBundle/Routing/DelegatingLoader.php]: Good catch. Fixed.

**Comment by Tobion** [comment]
Fixed dependencies. This will happen anyway eventually with all the types added in master.

File: src/Symfony/Bundle/FrameworkBundle/Command/AbstractConfigCommand.php [hunk 4]
```diff
@@ -52,6 +52,9 @@ protected function listBundles($output)
         }
     }
 
+    /**
+     * @return ExtensionInterface
```
Comment (nicolas-grekas) [inline src/Symfony/Bundle/FrameworkBundle/Command/AbstractConfigCommand.php]: that's for lower branches (all docblock changes on PRs on master should be on lower branches, ping @symfony/deciders and all other reviewers).

File: src/Symfony/Component/Config/Definition/Builder/ArrayNodeDefinition.php [hunk 5]
```diff
@@ -417,6 +411,10 @@ protected function createNode()
             }
 
             if ($this->default) {
+                if (!\is_array($this->defaultValue)) {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Config/Definition/Builder/ArrayNodeDefinition.php]: only for 5? missing deprecation in 4.4?

**Review by nicolas-grekas — COMMENTED** [review]
I'm good with the bump of lowest versions. Please move all updates to docblocks to a PR on a lower branch.

**Comment by Tobion** [comment]
> Please move all updates to docblocks to a PR on a lower branch.

I can do it for the single `@return ExtensionInterface` I added for autocomplete. But most other changes are about removing phpdocs that are obsolete with type hints. Those we should not apply in lower branches, because the type hints are not there yet.

File: src/Symfony/Component/Config/Definition/Builder/ArrayNodeDefinition.php (hunk 5 above)
Comment (Tobion) [reply src/Symfony/Component/Config/Definition/Builder/ArrayNodeDefinition.php]: I just moved the check from `PrototypedArrayNode.php::setDefaultValue` here so the typhint in `setDefaultValue` can be added. The `$this->defaultValue` comes from a generic setter that we cannot use types for.

File: src/Symfony/Component/Config/Definition/Builder/ArrayNodeDefinition.php [hunk 6]
```diff
@@ -417,6 +411,10 @@ protected function createNode()
...
+                    throw new \InvalidArgumentException($node->getPath().': the default value of an array node has to be an array.');
```
Comment (fabpot) [inline src/Symfony/Component/Config/Definition/Builder/ArrayNodeDefinition.php]: we use `sprintf()` in this case.

File: src/Symfony/Bundle/FrameworkBundle/Command/AbstractConfigCommand.php (hunk 4 above)
Comment (fabpot) [reply src/Symfony/Bundle/FrameworkBundle/Command/AbstractConfigCommand.php]: see #32382

File: src/Symfony/Component/Config/Definition/ArrayNode.php [hunk 7]
```diff
@@ -129,16 +129,16 @@ public function setAllowNewKeys(bool $allow)
      */
     public function setPerformDeepMerging(bool $boolean)
     {
-        $this->performDeepMerging = (bool) $boolean;
+        $this->performDeepMerging = $boolean;
     }
 
     /**
-     * Whether extra keys should just be ignore without an exception.
+     * Whether extra keys should just be ignored without an exception.
```
Comment (fabpot) [inline src/Symfony/Component/Config/Definition/ArrayNode.php]: #32382

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @Tobion.

File: src/Symfony/Component/Config/Definition/Builder/ArrayNodeDefinition.php (hunk 6 above)
Comment (fabpot) [reply src/Symfony/Component/Config/Definition/Builder/ArrayNodeDefinition.php]: fb3d92d799

File: src/Symfony/Bundle/FrameworkBundle/Command/AbstractConfigCommand.php (hunk 4 above)
Comment (Tobion) [reply src/Symfony/Bundle/FrameworkBundle/Command/AbstractConfigCommand.php]: Thanks @fabpot
